### DISCUSSION TITLE
Rez windows development workflow

### DISCUSSION THREAD

**Discussion by Jawabiscuit** [AcademySoftwareFoundation/rez#1301] [Q&A] [answered]
Yesterday I worked through getting setup for development on Windows using a virtual machine, Python3, virtualenv, and editable installs. I wrote down my experience and I'd like to consider this a good starting point but would like feedback from more seasoned devs. As an added effect, this discussion could be used by beginner and novice developers and enthusiasts as reference to make it easier to dive into tinkering or contributing to rez, until such time as an official "Best Practices" doc comes along.

In attempt to frame the conversation a bit more, some questions that might be answered, or open the door to answering, might include:

> As a developer, how do I get started developing rez?
> As an enthusiast, how do I get started debugging rez?
> As an enthusiast, how do I get started inspecting closer what's under the hood, maybe stepping through the code?
> As a developer, what's a setup for development that is efficient to get started with and applicable to a range of dev scenarios, long term as well as short?

Some backstory about me:

- a seasoned Python developer for Animation & VFX
- use rez+linux at work daily for development and testing dev releases
- use rez+windows at home daily for dev and environment management
- getting started with trying to debug/contribute to rez
- familiar but not as accustomed to following the 'normal' python developer workflow, e.g. setup.py
- familiar but generally don't use virtualenv for professional development
- haven't submitted a PR on github but many times on Bitbucket

Note: In anticipation of things changing I created [this gist](https://gist.github.com/Jawabiscuit/523b0d860a86304be0392f7cf9361975). I'll make any edits to that document, not the original text posted here.

My notes (copied from personal wiki):

## Rez  Getting Started

- [ ] I use a virtual machine that is setup using a [setup script](https://gist.github.com/Jawabiscuit/6298e2c4bebe5ddde694df9d18d9451b)
      - Python 2, 3 are installed but machine is configured (manually for now) so that Python 3 appears first in `*PATH` variables.
      - This guide uses `gitbash`. Powershell may be necessary given `gitbash` is still a new plugin for rez at the time of writing this.
- [ ] Install virtualenv
      - I prefer installing `virtualenv` and not using the builtin `venv` at the time of writing this.

`python -m pip install --user virtualenv`

![pip install virtualenv](https://bit.ly/3yb6t35)

- [ ] Add `virtualenv` executable to the `PATH`

`export PATH=$APPDATA/Python/Python310/Scripts:$PATH`

Or, for more consistent path delimiters:

```
export PATH="`cygpath -u "$APPDATA"`/Python/Python310/Scripts":$PATH
```

- [ ] Create a virtual env

`virtualenv venv`

![create virtualenv](https://bit.ly/3MthRLm)

- [ ] Install rez src in a virtualenv as an editable installation
      - I bootstrap this with a script depending on if there is other setup
      - A simple setup would just be `python -m pip install -e "git+https://github.com/Jawabiscuit/rez.git#egg=rez"`
      - This installs rez from my forked copy
      - Running `rez` inside the virtualenv executes `venv/Scripts/rez.exe`

```
# chdir to project
cd

# activate virtualenv
. venv/Scripts/activate

# editable install `-e`
`python -m pip install -e "git+https://github.com/Jawabiscuit/rez.git#egg=rez"`
```

![rez editable virtualenv install](https://bit.ly/3xVF7Oe)

- [ ] Warning: Pip-based rez installation detected

To my knowledge, production install is needed to circumnavigate this warning. See also: [Pip-based rez installation detected warning with installation script](https://github.com/nerdvegas/rez/discussions/1296)

![pip-based rez installation warning](https://bit.ly/3rPg9Mj)

- [ ] A `rez.egg-info` directory is installed in the `src` directory
      Todo: determine specifics of how this is used in python development

![rez.egg-info](https://bit.ly/3OFUShU)

- [ ] Create an editable production install

Where is the "production install"?

![rez production install](https://bit.ly/3vNhkNw)

- [ ] Edit pip install command in rez src
      - It may be easiest to use vscode.dev on a virtual machine.

Based on this [Slack discussion](https://academysoftwarefdn.slack.com/archives/C0321B828FM/p1650655464411449)

![slack discussion](https://bit.ly/3y0fHPz)

- [ ] Create dev branch
      - Note usage of `""` in the screenshot is actually a typo, it isn't necessary

`git checkout -b production-editable-install`

![git branch](https://bit.ly/3rTDLiV)

- [ ] Edit [install.py](https://github.com/Jawabiscuit/rez/commit/5891bbbec6a0ab9827f7987ee75fcbaad8de1353) (link shows the edit)

![edit install.py](https://bit.ly/3rOZOHC)

- [ ] Install and verify production install uses edited pip command

`python install.py -v /c/rez`

![run install.py](https://bit.ly/3rUDg8l)

- [ ] If using a virtual machine, save a snapshot after installation.

![vm snapshot](https://bit.ly/3MuLgVw)

- [ ] Add production install rez paths, either in a new terminal or in bash rc

```bash
export PATH=/c/rez/Scripts/rez:$PATH

export PYTHONPATH=/c/rez/Lib/site-packages:/c/rez/Scripts

rez --version
```

- [ ] Production install rez directory structure

![production rez directory](https://bit.ly/3vRxLJ0)

- [ ] Add venv directory to editor workspace

- [ ] Insert [print message](https://github.com/Jawabiscuit/rez/commit/5891bbbec6a0ab9827f7987ee75fcbaad8de1353) in `_main.py` (link shows the edit)

- [ ] Print from rez cli
      - Open a new terminal, if not done so already
      - Test print message using the randomly chosen `rez plugins` command, or other command depending on where the message is inserted.

![print message](https://bit.ly/3OEAxJT)

- [ ] Rez is now installed to a production install location as an 'editable' install as well as uses source code that you can edit in your virtualenv!

## Reference

[pip editable installs](https://pip.pypa.io/en/stable/cli/pip_install/?highlight=editable#editable-installs)
[setuptools development](https://setuptools.pypa.io/en/latest/userguide/development_mode.html)
[how do editable pip installs work? anthony explains](https://www.youtube.com/watch?v=gYYi7varbmE)

**Comment by ColinKennedy** [comment]
I was just asking about editable pip installed earlier so this post came at a great time. Thank you for explaining it in detail!

**Reply by Jawabiscuit** [reply]
Recent discussions motivated me ;) I linked that slack thread in the post.

**Comment by Jawabiscuit** [comment, marked as answer]
I've given the gist some more TLC after testing the workflow on another Windows machine. As disclosed in the original post, the above documentation is out now out of date but will remain for posterity. The original idea was to get some feedback from the community at the time of creating this discussion. Anyone reading this should refer to the [gist](https://gist.github.com/Jawabiscuit/523b0d860a86304be0392f7cf9361975) for the latest information.
